CORDA: Store signatures before sending transactions to Notary

yerinle opened the pull request:
Store signatures before sending transactions to Notary

yerinle committed:
ENT-6875: Update FinalityFlow to record transaction locally before sending to notary

yerinle committed:
ENT-6875: Update RecieveFinalityFlow to access second subFlow call

yerinle committed:
ENT-6875: Add signatures and notary_status columns to NODE_TRANSACTIONS table

yerinle committed:
ENT-6875: Remove call in DBTransactionsResolver

yerinle committed:
ENT-6875: Use another function to store signatures

yerinle committed:
ENT-6875: Update notary status column when adding verified transaction

yerinle committed:
ENT-6875: Some cleanup

blsemo commented on core/src/main/kotlin/net/corda/core/flows/FinalityFlow.kt:
```diff
@@ -172,15 +175,33 @@ class FinalityFlow private constructor(val transaction: SignedTransaction,
-
         progressTracker.currentStep = BROADCASTING
 
+        logger.info("About to broadcast transaction to other parties (minus Notary signature)")
```
Surely, these can't be logged at info level when merging to release branch.

yerinle replied:
changed to debug

blsemo commented on core/src/main/kotlin/net/corda/core/flows/FinalityFlow.kt:
```diff
@@ -282,13 +333,18 @@ class ReceiveFinalityFlow @JvmOverloads constructor(private val otherSideSession
-                            "not recorded and nor its states sent to the vault."
-                }
-            }
+        logger.info("1st ReceiveFinalityFlow subFlow called")
```
These log messages should state what is actually happening (i.e. "Waiting to receive transaction" "Waiting to receive notary signatures") - whether that is the first or second subflow is irrelevant for someone reading the logs.

yerinle replied:
apologies removed, added for debugging purposes

blsemo commented on core/src/main/kotlin/net/corda/core/flows/FinalityFlow.kt:
```diff
@@ -172,15 +175,33 @@ class FinalityFlow private constructor(val transaction: SignedTransaction,
+        logger.info("About to broadcast transaction to other parties (minus Notary signature)")
+        broadcastToOtherParties(externalTxParticipants, transaction)
+        logger.info("All parties received the transaction successfully. (minus Notary signature)")
+        recordSignaturesLocally(transaction)
```
I don't quite understand that logic. Wouldn't we want to record a copy of the TX before trying to notarise?

yerinle replied:
yes that's correct. This page details the point of these changes https://r3-cev.atlassian.net/wiki/spaces/CB/pages/3821436942/Corda+4+Customer+Feedback+Solution+Ideas   check the "More Reliable Finality / Consensus" section

blsemo commented on core/src/main/kotlin/net/corda/core/flows/FinalityFlow.kt:
```diff
@@ -172,15 +175,33 @@ class FinalityFlow private constructor(val transaction: SignedTransaction,
+            val notarisedSigs = notarised.sigs - transaction.sigs.toSet()
+            broadcastToOtherParties(notarisedSigs)
+            logger.info("All parties received the extra signatures successfully.")
+            recordTransactionLocally(notarised)
+        } else {
+            recordTransactionLocally(transaction)
```
Why would we not always record `notarised`?

yerinle replied:
my understanding is certain transactions don't get notarized

blsemo commented on core/src/main/kotlin/net/corda/core/flows/FinalityFlow.kt:
```diff
@@ -192,12 +213,26 @@ class FinalityFlow private constructor(val transaction: SignedTransaction,
+    }
+    @Suspendable
+    private fun broadcastToOtherParties(sigs: List<TransactionSignature>) {
+        if (newApi) {
```
So what would happen if not using the new API or with old nodes? If I read the code correctly, they would receive the tx pre notarisation, but never get sent the notary signature?

blsemo commented on node/src/main/kotlin/net/corda/node/services/persistence/DBTransactionStorage.kt:
```diff
@@ -256,6 +340,20 @@ class DBTransactionStorage(private val database: CordaPersistence, cacheFactory:
         }
     }
 
+    override fun addNotNotarizedTransactionSigs(transaction: SignedTransaction) {
```
That function seems misnamed - it is adding/storing the SignedTransaction, not just the Signatures.

yerinle committed:
ENT-6875: Update log levels and cleanup

yerinle committed:
ENT-6875: making changes from code review

adelel1 commented on node/src/main/kotlin/net/corda/node/services/persistence/DBTransactionStorage.kt:
```diff
@@ -270,12 +387,18 @@ class DBTransactionStorage(private val database: CordaPersistence, cacheFactory:
                 stx.txBits,
                 Collections.unmodifiableList(stx.sigs),
                 status)
+        constructor(stx: SignedTransaction, status: TransactionStatus, notaryStatus: NotaryStatus) : this(
```
We need the additional signatures loaded here. NotaryStatus can be combined into Transaction Status.

yerinle committed:
ENT-6875: Retrieve notary sig when reading transaction

yerinle committed:
ENT-6875: Clean-up and adding more tests

adelel1 commented on core/src/main/kotlin/net/corda/core/flows/FinalityFlow.kt:
```diff
@@ -282,13 +333,16 @@ class ReceiveFinalityFlow @JvmOverloads constructor(private val otherSideSession
-                            "not recorded and nor its states sent to the vault."
-                }
-            }
+        val stx = subFlow(object : ReceiveTransactionFlow(otherSideSession, checkSufficientSignatures = false, statesToRecord = statesToRecord) {
```
Rick pointed out here that there is a side effect of the checkSufficientSignatures being false, the txn is not recorded in this case.
Suggestion here is to have two paths for the code i.e.....

if (needsNotarySignature ) {
subFlow....
}
else {
// Notary sig not needed so checkSufficientSignatures can be true and txn will be recorded
subFlow....
}

Also I noticed the checkBeforeRecording callback has been removed as well? Why? Can it be inserted back in?

adelel1 commented:
The code needs to be compatible with older nodes. So in the FinalityFlow call method we need to know the version running on the counterparty.  We can use platform version for this. This is available on the network map. Which you can get from the serviceHub which is accessible. So if platform version is 13 or greater on the counterparty you know its running the new code, so we can communicate using the new message sequence. Otherwise maintain the old message sequence.

In enterprise the communications are done in parallel when sending to the other nodes. So we figured you can partition the counterparty sessions into two sets, one for old api and for new api. We can do this in OS as well.

Also dotted around the code there is a newApi boolean, this is to distinguish between Corda 4 (new) and Corda 3 old. So we can keep existing code as is if newApi = false (for Corda 3)

Some pseudo code

if (!newApi) {
   // this is Corda 3 code, we can keep all this the same
}
else {
    // We are running C4 code
    // Now need to partition sessions into 2 sets, one for platform version >=13 and one for platform version <13

    for all sessions where counterparty has platform version >=13 do {
        // use new code
    }
    for all sessions where counterparty has platform version <13 do {
       // use existing code
   }
}

Note currently for 4.10 platform version is 12, for 4.11 platform version will be 13.

adelel1 commented on node/src/main/kotlin/net/corda/node/services/persistence/DBTransactionStorage.kt:
```diff
@@ -180,6 +190,57 @@ class DBTransactionStorage(private val database: CordaPersistence, cacheFactory:
+                val criteriaUpdate = criteriaBuilder.createCriteriaUpdate(DBTransaction::class.java)
+                val updateRoot = criteriaUpdate.from(DBTransaction::class.java)
+                criteriaUpdate.set(updateRoot.get<ByteArray>(DBTransaction::signatures.name), signatures.serialize(context = contextToUse().withEncoding(SNAPPY)).bytes)
+                criteriaUpdate.where(criteriaBuilder.and(
```
Should check here the current status of the transaction, so that we don't find ourselves overwriting signatures that are already verified. i.e. only record if status is missing notary signature. 
We could have an extra parameter here passed in, which when set will flip the verified flag of the transaction as well.

yerinle committed:
ENT-6875: Make changes backwards compatible with older platform versions

adelel1 commented on core/src/main/kotlin/net/corda/core/flows/FinalityFlow.kt:
```diff
@@ -172,15 +175,47 @@ class FinalityFlow private constructor(val transaction: SignedTransaction,
+
+        val (lessThan13, equalGreater13) = sessions.partition { serviceHub.networkMapCache.getNodeByLegalName(it.counterparty.name)?.platformVersion!! < 13 }
+
+        if (platformVersion >= 13) {
```
Note that this platformVersion is our platform version, which will always be 13 or greater. So this statement will always be true. 
For old counter parties we can also use new db format. i.e. Store txn first without notary sig, get the notary sig then store extra sigs. The difference is what is broadcasted to other parties. 
So its the broadcastToOtherParties(externalTxParticipants, equalGreater13, transaction), that will only be done if equalGreater13 is not empty.

adelel1 commented on core/src/main/kotlin/net/corda/core/flows/FinalityFlow.kt:
```diff
@@ -172,15 +175,47 @@ class FinalityFlow private constructor(val transaction: SignedTransaction,
         val notarised = notariseAndRecord()
 
-        progressTracker.currentStep = BROADCASTING
+        if (platformVersion >= 13) {
```
Comment applies to notariseAndRecord. Now though the notariseAndRecord is not doing any recording. Can you update the func names to reflect what they are doing (at least the new/updated ones). I know in the past this was a problem as well.

adelel1 commented on core/src/main/kotlin/net/corda/core/flows/FinalityFlow.kt:
```diff
@@ -172,15 +175,47 @@ class FinalityFlow private constructor(val transaction: SignedTransaction,
 
+            return notarised
+        } else {
+            broadcastToOtherParties(externalTxParticipants, lessThan13, notarised)
```
Note for old code the txn has not been recorded, as notariseAndRecord has changed.

adelel1 commented on core/src/main/kotlin/net/corda/core/flows/FinalityFlow.kt:
```diff
@@ -192,12 +227,26 @@ class FinalityFlow private constructor(val transaction: SignedTransaction,
+    }
+    @Suspendable
+    private fun broadcastToOtherParties(sigs: List<TransactionSignature>) {
+        if (newApi) {
```
If (newApi) means if Corda4, so don't need this check for any new code.

yerinle committed:
ENT-6875: function renaming and general tidy-up

yerinle committed:
Merge branch 'release/os/4.10' into yinka/ENT-6875

# Conflicts:
#	constants.properties
#	core/src/main/kotlin/net/corda/core/flows/FinalityFlow.kt
#	core/src/main/kotlin/net/corda/core/internal/CordaUtils.kt

yerinle committed:
ENT-6875: Updating FinalityFlow after merge

rick-r3 commented on core/src/main/kotlin/net/corda/core/flows/FinalityFlow.kt:
```diff
@@ -173,15 +176,38 @@ class FinalityFlow private constructor(val transaction: SignedTransaction,
+    private fun broadcastToOtherParties(externalTxParticipants: Set<Party>, sessions: Collection<FlowSession>, tx: SignedTransaction) {
         if (newApi) {
-            oldV3Broadcast(notarised, oldParticipants.toSet())
+            oldV3Broadcast(tx, oldParticipants.toSet())
```
I think we should only be doing the old V3 broadcast when this is the final `broadcastToOtherParties()` with the fully signed transaction (line 199).  Add a boolean flag to the args to the method and conditionally call `oldV3Broadcast...` , with the call at line 185 passing `false` and line 199 passing `true` ?

rick-r3 replied:
name of the flag could be `transactionFullySigned`

colljos replied:
Addressed

rick-r3 commented on node/src/main/kotlin/net/corda/node/services/persistence/DBTransactionStorage.kt:
```diff
@@ -276,7 +352,10 @@ class DBTransactionStorage(private val database: CordaPersistence, cacheFactory:
-
+        constructor(stx: SignedTransaction, sigs: List<TransactionSignature>?, status: TransactionStatus) : this(
+                stx.txBits,
+                if (sigs == null) Collections.unmodifiableList(stx.sigs) else Collections.unmodifiableList(stx.sigs) + sigs,
```
Just to be sure it `unmodifiable`, I would put the "+" inside parentheses e.g. `Collections.unmodifiableList(stx.sigs + sigs)`

rick-r3 reviewed (commented):
Some overall comments:

1. Need to ack back after initial transaction (without notary sig) is broadcast & recorded on remote peers - otherwise can go to the notary before any of the other parties have actually recorded the transaction - which is one of the problems we are trying to address here.
2. Where we have a broadcast & ack pattern, better to split the sends and receives as we’ll want this to be performant in enterprise edition.  My recommendation is to broadcast -> record local -> wait for ack, so that local recording happens in parallel with remote recording.  record local -> broadcast -> wait for ack is also valid but will perform worse as record local and record remote will happen sequentially.
3. Combine recording of additional (notary) sigs and flipping the status to verified into single update statement.  Obviously going to be more efficient and we care about performance eventually, and it changes the structure of underlying DBTransactionStorage methods and gives us less race conditions etc.
4. I think DBTransactionStorage.addTransaction() is getting really overloaded with (legacy) use cases which makes optimising things when we look at enterprise later hard to reason about, and probably lead to the two stage processing of adding sigs and then flipping the status I mention in the last point.  I would probably turn ServiceHub.recordExtraSignatures() into a ServiceHub.finalizeTransactionWithExtraSignatures() that shares a lot of logic with ServiceHub.recordTransactions() (refactor that method to have the bulk of the work done by a private implementation that takes a lambda where it currently does “validatedTransactions::addTransaction”) and utilise a single method on DBTransaction storage to store the signatures and flip the status.
5. I think the “finalise…” method/code-path should only update the row if the row status is “missing sigs”.
6.  When ack'ing I recommend sending "Unit" which is a special kotlin singleton type with no value (sort of void-like), rather than a string.  We've done that elsewhere I think.

A few specific comments elsewhere - particularly around compatibility.  I've skipped some detailed comments until the high level layout is closer as they might well be redundant then.

yerinle committed:
ENT-6875: Fixing code check and test failures

yerinle committed:
ENT-6875: Code review feedback

rick-r3 commented on core/src/main/kotlin/net/corda/core/flows/FinalityFlow.kt:
```diff
@@ -178,38 +179,37 @@ class FinalityFlow private constructor(val transaction: SignedTransaction,
+    private fun broadcastToOtherParties(externalTxParticipants: Set<Party>, sessions: Collection<FlowSession>, tx: SignedTransaction, transactionFullySigned: Boolean) {
         if (newApi) {
-            oldV3Broadcast(tx, oldParticipants.toSet())
+            if (transactionFullySigned) oldV3Broadcast(tx, oldParticipants.toSet())
```
Think you also need this condition at the other call site for `oldV3Broadcast` in the `else` branch below.

rick-r3 commented on core/src/main/kotlin/net/corda/core/flows/FinalityFlow.kt:
```diff
@@ -357,13 +357,13 @@ class ReceiveFinalityFlow @JvmOverloads constructor(private val otherSideSession
                     .unwrap { it }
             serviceHub.recordExtraSignatures(stx.id, signatures)
             serviceHub.recordTransactions(statesToRecord, setOf(stx))
-            otherSideSession.send("Received signatures")
```
You need a `send(Unit)` in this `if` branch after `recordTransactionWithoutNotarySignature()` - it needs to be before the `receive` of the sigs - and then one in the `else` branch below after `recordTransactions()`.

The one in the `else` branch might be talking to an "old" node that isn't expecting the `send`, so you need to conditionally do it based on the platform version of `otherSideSession` (do similar lookup as you did before to generate old and new session lists).

rick-r3 commented on core/src/main/kotlin/net/corda/core/flows/FinalityFlow.kt:
```diff
@@ -228,7 +228,7 @@ class FinalityFlow private constructor(val transaction: SignedTransaction,
         for (session in sessions) {
             try {
-                session.sendAndReceive<String>(sigs)
+                session.sendAndReceive<Unit>(sigs)
```
I think this just becomes a `send(sigs)` only - nothing to receive.

rick-r3 commented on core/src/main/kotlin/net/corda/core/flows/FinalityFlow.kt:
```diff
@@ -178,38 +179,37 @@ class FinalityFlow private constructor(val transaction: SignedTransaction,
             for (session in sessions) {
                 try {
                     subFlow(SendTransactionFlow(session, tx))
+                    session.receive<Unit>()
```
When `transactionFullySigned` is `true`, you are talking to an old node, so don't do this `receive` when `true`.

rick-r3 reviewed (commented):
Comment on commit b4d7542027a7bceed3dd23e7f3346964eaa88b63

yerinle committed:
ENT-6875: Code review feedback

rick-r3 commented on core/src/main/kotlin/net/corda/core/flows/FinalityFlow.kt:
```diff
@@ -287,13 +343,28 @@ class ReceiveFinalityFlow @JvmOverloads constructor(private val otherSideSession
             }
         })
+        if(needsNotarySignature(stx)) {
+            serviceHub.recordTransactionWithoutNotarySignature(setOf(stx))
```
I _think_ this flow is basically correct now

colljos replied:
I think we need some conditional logic to only expect signatures from peers that are running the 2_PHASE_FINALITY flow (eg. platformVersion >= 13).

rick-r3 commented on core/src/main/kotlin/net/corda/core/flows/FinalityFlow.kt:
```diff
@@ -173,15 +177,39 @@ class FinalityFlow private constructor(val transaction: SignedTransaction,
+            serviceHub.finalizeTransactionWithExtraSignatures(ONLY_RELEVANT, listOf(notarised), notarisedSigs)
+        }
+
+        recordTransactionLocally(notarised)
```
I think after recording, you need to do a `receive<Unit>()` from each of the `newPlatformSessions` to match the conditional `send(Unit)` at the end of `ReceiveFinalityFlow`

rick-r3 commented on core/src/main/kotlin/net/corda/core/flows/FinalityFlow.kt:
```diff
@@ -173,15 +177,39 @@ class FinalityFlow private constructor(val transaction: SignedTransaction,
+            logger.debug("About to broadcast extra signatures to other parties")
+            broadcastToOtherParties(notarisedSigs, newPlatformSessions)
+            logger.debug("All parties received the extra signatures successfully.")
+            serviceHub.finalizeTransactionWithExtraSignatures(ONLY_RELEVANT, listOf(notarised), notarisedSigs)
```
Replace `ONLY_RELEVANT` here with `statesToRecord`

rick-r3 replied:
I think rather than `listOf(notarised)` use `listOf(transaction)` as otherwise you sort of supply the extra sigs twice, since you already pass them in the extra `notarisedSigs` arg.

colljos replied:
Need to finalize with notary signed transaction.

rick-r3 commented on core/src/main/kotlin/net/corda/core/flows/FinalityFlow.kt:
```diff
@@ -173,15 +177,39 @@ class FinalityFlow private constructor(val transaction: SignedTransaction,
+        val notarised = notariseIfNotarySignatureIsRequired()
+
+        if (notarised.sigs != transaction.sigs) {
+            val notarisedSigs = notarised.sigs - transaction.sigs.toSet()
```
Invert this and the line above so you have 

```
val notarisedSigs = notarised.sigs - transaction.sigs.toSet()
if(notarisedSigs.isNotEmpty()) {
```

yerinle replied:
Made requested changes

rick-r3 reviewed (commented):
Some more comments.  Think this should give roughly the right shape to the flows, and cope with compatibility.

yerinle committed:
ENT-6875: Code Review Feedback

yerinle committed:
ENT-6875: Ignore some failing tests see if builds succeeds

yerinle committed:
ENT-6875: Ignore some failing tests see if builds succeeds

yerinle committed:
ENT-6875: Ignore failing integration tests for now

yerinle committed:
ENT-6875: Ignore failing integration tests for now

yerinle committed:
ENT-6875: Suppress invalid code checks

yerinle committed:
ENT-6875: Fix and un-ignore some tests

yerinle committed:
ENT-6875: Remove unused imports

yerinle committed:
ENT-6875: Ignore some failing integration tests for now

yerinle committed:
ENT-6875: Ignore some failing integration tests for now

colljos committed:
Add missing implementations of new ServiceHub functions.
Remove @Ignored tests.

colljos committed:
Merge remote-tracking branch 'origin/release/os/4.10' into yinka/ENT-6875

colljos commented:
> 1. Need to ack back after initial transaction (without notary sig) is broadcast & recorded on remote peers - otherwise can go to the notary before any of the other parties have actually recorded the transaction - which is one of the problems we are trying to address here.

Addressed by Yinka.

> 2. Where we have a broadcast & ack pattern, better to split the sends and receives as we’ll want this to be performant in enterprise edition.  My recommendation is to broadcast -> record local -> wait for ack, so that local recording happens in parallel with remote recording.  record local -> broadcast -> wait for ack is also valid but will perform worse as record local and record remote will happen sequentially.

JC: For the second broadcast (after notarisation), the danger here is that we finalize the transaction locally before all the remote participants have acknowledged receipt of notary signatures. wdyt @rick-r3 ?

> 3. Combine recording of additional (notary) sigs and flipping the status to verified into single update statement.  Obviously going to be more efficient and we care about performance eventually, and it changes the structure of underlying DBTransactionStorage methods and gives us less race conditions etc.

Addressed by Yinka.

> 4. I think DBTransactionStorage.addTransaction() is getting really overloaded with (legacy) use cases which makes optimising things when we look at enterprise later hard to reason about, and probably lead to the two stage processing of adding sigs and then flipping the status I mention in the last point.  I would probably turn ServiceHub.recordExtraSignatures() into a ServiceHub.finalizeTransactionWithExtraSignatures() that shares a lot of logic with ServiceHub.recordTransactions() (refactor that method to have the bulk of the work done by a private implementation that takes a lambda where it currently does “validatedTransactions::addTransaction”) and utilise a single method on DBTransaction storage to store the signatures and flip the status.

Addressed by JC.

> 5. I think the “finalise…” method/code-path should only update the row if the row status is “missing sigs”.

Addressed by JC.

> 6. When ack'ing I recommend sending "Unit" which is a special kotlin singleton type with no value (sort of void-like), rather than a string.  We've done that elsewhere I think.

Addressed by Yinka.

colljos committed:
Incorporating PR review feedback from RP comments:
1. Where we have a broadcast & ack pattern, better to split the sends and receives as we’ll want this to be performant in enterprise edition. My recommendation is to broadcast -> record local -> wait for ack, so that local recording happens in parallel with remote recording.
2. I would probably turn ServiceHub.recordExtraSignatures() into a ServiceHub.finalizeTransactionWithExtraSignatures() that shares a lot of logic with ServiceHub.recordTransactions() (refactor that method to have the bulk of the work done by a private implementation that takes a lambda where it currently does “validatedTransactions::addTransaction”)

colljos committed:
Fix detekt complaint.

colljos committed:
Fixed failing FlowHospitalTests.

colljos committed:
Fixed failing FlowHospitalTests - clean-up.

colljos committed:
Re-instate notariseAndRecord.

colljos committed:
Fix failing tests.

colljos committed:
Fix detekt failure.

colljos committed:
Fix failing CertificateRotationTest test.

colljos committed:
Fix VaultObserverExceptionTest failing test.
Optimise-out final Ack post finalisation (TBC)

colljos committed:
Remove unused imports.

colljos committed:
Fix failing FlowReloadAfterCheckpointTest.kt test.

colljos committed:
Finality Flow processing based on runtime platform version.
Added test cases to validate mixed initiator-receiver finality flow scenarios.

colljos committed:
Fix failing test in ReferencedStatesFlowTests

colljos commented:
Superseded by https://github.com/corda/corda/pull/7290
